{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import print_function\n",
    "import vcf \n",
    "import sierrapy\n",
    "import json\n",
    "import requests\n",
    "import subprocess\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [],
   "source": [
    "aminoAcids ={   \n",
    "#Phenylalanine    \n",
    "\"TTT\": \"F\",\n",
    "\"TTC\": \"F\",\n",
    "#Leucine    \n",
    "\"TTA\": \"L\",     \n",
    "\"TTG\": \"L\",\n",
    "\"CTT\": \"L\",\n",
    "\"CTC\": \"L\",\n",
    "\"CTA\": \"L\",\n",
    "\"CTG\": \"L\", \n",
    "#Isoleucine\n",
    "\"ATT\": \"I\",\n",
    "\"ATC\": \"I\", \n",
    "\"ATA\": \"I\", \n",
    "#Methioinine    \n",
    "\"ATG\": \"M\", \n",
    "#Valine  \n",
    "\"GTT\": \"V\", \n",
    "\"GTC\": \"V\",\n",
    "\"GTA\": \"V\",\n",
    "\"GTG\": \"V\",\n",
    "#Serine\n",
    "\"TCT\": \"S\",\n",
    "\"TCC\": \"S\",\n",
    "\"TCA\": \"S\",\n",
    "\"TCG\": \"S\",\n",
    "\"AGT\": \"S\",\n",
    "\"AGC\": \"S\",    \n",
    "#Proline  \n",
    "\"CCT\": \"P\",\n",
    "\"CCC\": \"P\",\n",
    "\"CCA\": \"P\",\n",
    "\"CCG\": \"P\",\n",
    "#Threonine\n",
    "\"ACT\": \"T\",\n",
    "\"ACC\": \"T\",\n",
    "\"ACA\": \"T\",\n",
    "\"ACG\": \"T\",      \n",
    "#Alanine\n",
    "\"GCT\": \"A\",\n",
    "\"GCC\": \"A\",\n",
    "\"GCA\": \"A\",\n",
    "\"GCG\": \"A\",\n",
    "#Tyrosine\n",
    "\"TAT\": \"Y\",\n",
    "\"TAC\": \"Y\",\n",
    "#Histidine\n",
    "\"CAT\": \"H\",\n",
    "\"CAC\": \"H\",\n",
    "#Glutamine\n",
    "\"CAA\": \"Q\",\n",
    "\"CAG\": \"Q\",\n",
    "#Asparagine\n",
    "\"AAT\": \"N\",\n",
    "\"AAC\": \"N\",\n",
    "#Lysine\n",
    "\"AAA\": \"K\",\n",
    "\"AAG\": \"K\",\n",
    "#Aspartic acid\n",
    "\"GAT\": \"D\",\n",
    "\"GAC\": \"D\",\n",
    "#Glutamic acid\n",
    "\"GAA\": \"E\",    \n",
    "\"GAG\": \"E\",\n",
    "#Cysteine\n",
    "\"TGT\": \"C\",\n",
    "\"TGC\": \"C\",\n",
    "#Tryptophan\n",
    "\"TGG\": \"W\",\n",
    "#Arginine\n",
    "\"CGT\": \"R\",\n",
    "\"CGC\": \"R\",\n",
    "\"CGA\": \"R\",\n",
    "\"CGG\": \"R\", \n",
    "\"AGA\": \"R\",\n",
    "\"AGG\": \"R\",     \n",
    "#Glycine\n",
    "\"GGT\": \"G\",\n",
    "\"GGC\": \"G\",\n",
    "\"GGA\": \"G\",\n",
    "\"GGG\": \"G\",    \n",
    "#stop\n",
    "\"TAA\": \"*\",\n",
    "\"TAG\": \"*\",  \n",
    "\"TGA\": \"*\"         \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getCharacter(position):\n",
    "    fp = open('sequence.fasta', 'r')\n",
    "    line_num = int(position//70 + 1) # add one to skip the first line\n",
    "    char_num = int(position%70)\n",
    "    print(\"Line Number: \", line_num, \"Character Number: \", char_num)\n",
    "    for i, line in enumerate(fp):\n",
    "        if i == line_num:\n",
    "            return line[char_num]\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getCodon(start,position,alt):\n",
    "    adjusted = position - start\n",
    "    print(\"Adjusted\",adjusted)\n",
    "    #start\n",
    "    if adjusted%3 == 0:\n",
    "        print(\"start\")\n",
    "        codon.append(alt)#codon.append(getCharacter(position))\n",
    "        codon.append(getCharacter(position+1))\n",
    "        codon.append(getCharacter(position+2))\n",
    "    #middle\n",
    "    elif adjusted%3 == 1:\n",
    "        print(\"middle\")\n",
    "        codon.append(getCharacter(position-1))\n",
    "        codon.append(alt)#codon.append(getCharacter(position))\n",
    "        codon.append(getCharacter(position+1))\n",
    "    #end\n",
    "    elif adjusted%3 ==2:\n",
    "        print(\"end\")\n",
    "        codon.append(getCharacter(position-2))\n",
    "        codon.append(getCharacter(position-1))\n",
    "        codon.append(alt)#codon.append(getCharacter(position))\n",
    "    codon.append(adjusted+1) #position number in protease region. starts at 1\n",
    "    return codon   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getSequence(start,end):\n",
    "    fp = open('sequence.fasta', 'r')\n",
    "    print(\"Get Sequence\")\n",
    "    start_line = start//70 + 1\n",
    "    end_line = end//70 + 1\n",
    "    char_start = start%70\n",
    "    char_end = end%70\n",
    "    sequence = \"\"\n",
    "    #print(start_line, end_line, char_start, char_end)\n",
    "    for i, line in enumerate(fp):\n",
    "        if i >= start_line and i <= end_line:\n",
    "            if i == start_line:\n",
    "                sequence = line[char_start:70]\n",
    "            elif i == end_line:\n",
    "                sequence = sequence + line[0:char_end+1]\n",
    "            else:\n",
    "                sequence = sequence + line[0:70]\n",
    "            #print(sequence,char_end)\n",
    "    fp.close()\n",
    "    return sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getGeneRegions():  \n",
    "    \n",
    "    #protease\n",
    "    pStart = 2252\n",
    "    pEnd = 2548\n",
    "    protease = getSequence(pStart, pEnd)\n",
    "    pf = open(\"protease.txt\", \"w\")\n",
    "    pf.write(protease)\n",
    "    pf.close()\n",
    "    #reverse transcriptase\n",
    "    rtStart = 2549\n",
    "    rtEnd = 4228 #3868 including rnase\n",
    "    rTranscriptase = getSequence(rtStart, rtEnd)\n",
    "    rf = open(\"reverse_transcriptase.txt\", \"w\")\n",
    "    rf.write(rTranscriptase)\n",
    "    rf.close()\n",
    "    iStart = 4229\n",
    "    iEnd = 5092 #5095 stop codon is not included\n",
    "    integrase = getSequence(iStart, iEnd)\n",
    "    intf = open(\"integrase.txt\", \"w\")\n",
    "    intf.write(integrase)\n",
    "    intf.close()\n",
    "  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getAminoAcids(inputF,outputF):\n",
    "\n",
    "    inputFile = open(inputF, 'r')\n",
    "    translation = \"\"\n",
    "    sequence = \"\"\n",
    "    outputFile= open(outputF, \"w\")\n",
    "    for i, line in enumerate(inputFile):\n",
    "        sequence = sequence + line\n",
    "    seq_len = len(sequence)\n",
    "    print(\"length of sequence\",seq_len)\n",
    "    for i in range(0,seq_len,3):\n",
    "         #print(\"seq\",sequence[i:i+3])\n",
    "        translation = translation + aminoAcids.get(sequence[i:i+3])\n",
    "    print(\"translated\",translation)\n",
    "    outputFile.write(translation)\n",
    "    inputFile.close()\n",
    "    outputFile.close()\n",
    "    return translation\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getCombinations(base1,  base2, base3,reference):\n",
    "# returns all possible combinations \n",
    " \n",
    "    mutations = []\n",
    "    #print(\"Bases\", base1, base2, base3)\n",
    "    for i in range (len(base1)):\n",
    "        for j in range (len(base2)):\n",
    "            for k in range (len(base3)):\n",
    "                codonStr = str(base1[i])+str(base2[j])+str(base3[k])\n",
    "                #print(\"CODON:\",codonStr)\n",
    "                trans = aminoAcids.get(codonStr,\"!\")\n",
    "                #print(\"TRANS:\",trans, reference)\n",
    "                if trans!=\"!\" and trans!=reference: \n",
    "                    #add as a mutation only if it differs from the reference\n",
    "                    if trans not in mutations:\n",
    "                        #print(\"ALT codon\",codonStr)\n",
    "                        mutations.append(trans)\n",
    "    return mutations        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getMutations_mult(mut,refG): \n",
    "    #returns all possible amino acid mutations per position\n",
    "    #protease\n",
    "    mutList = []\n",
    "    geneF = open(refG, 'r')\n",
    "    ref_gene = [ch for ch in geneF.read()]\n",
    "    codonct = 1\n",
    "    #print(\"LENPROT\",len(prot))\n",
    "    for i in range(0,len(prot),3):\n",
    "        #print(\"I\",i,prot[i])\n",
    "        #print(\"CODONCT \",codonct-1,ref_protease[codonct-1])\n",
    "        #print(\"Base:\",prot[i],prot[i+1],prot[i+2])\n",
    "        combinations = getCombinations(mut[i],mut[i+1], mut[i+2],ref_gene[codonct-1])\n",
    "        if len(combinations)>0:\n",
    "            print(\"MUTATION:\", codonct, ref_gene[codonct-1], \"comb\", combinations)\n",
    "            mutList.append([codonct,ref_gene[codonct-1],combinations])\n",
    "           \n",
    "        codonct = codonct + 1\n",
    "        \n",
    "    print(\"M\",mutList) \n",
    "    return mutList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [],
   "source": [
    "def insertVariants_mult(vcfFile,pFileName, rtFileName, iFileName): \n",
    "    #gets all possible alternative bases per position\n",
    "    vcf_reader = vcf.Reader(open(vcfFile,'r'))  \n",
    "    pFile = open(pFileName, 'r')\n",
    "    protease_seq = [ch for ch in pFile.read()]\n",
    "    rtFile = open(rtFileName, 'r')\n",
    "    rt_seq = [ch for ch in rtFile.read()]   \n",
    "    iFile = open(iFileName, 'r')\n",
    "    i_seq = [ch for ch in iFile.read()]\n",
    "    #protease\n",
    "    for record in vcf_reader:\n",
    "       \n",
    "        if record.POS >=2253 and record.POS <2550: #protease\n",
    "            #print(\"variant\",record.POS,record.POS-2253,record.REF,record.ALT)\n",
    "            #print(record)\n",
    "            alt = record.ALT\n",
    "            \n",
    "            temp = []\n",
    "            #temp.append(record.REF)\n",
    "            #temp.append(protease_seq[record.POS-2253])\n",
    "            for i in range(len(alt[0])):\n",
    "                base = str(alt[0])[i]\n",
    "                if base not in temp:\n",
    "                    temp.append(base)\n",
    "            protease_seq[record.POS-2253] = temp \n",
    "            #print( \"prot\",protease_seq[record.POS-2253])\n",
    "        elif record.POS >=2550 and record.POS < 4229: #3869: #reverse transcriptase\n",
    "            #print(\"variant\",record.POS,record.POS-2550,record.REF,record.ALT)\n",
    "            alt = record.ALT\n",
    "            temp = []\n",
    "           #temp.append(record.REF)\n",
    "            #temp.append(rt_seq[record.POS-2550])\n",
    "            for i in range(len(alt[0])):\n",
    "                base = str(alt[0])[i]\n",
    "                if base not in temp:\n",
    "                    temp.append(base)\n",
    "            rt_seq[record.POS-2550] = temp \n",
    "            #print(rt_seq[record.POS-2550])\n",
    "        elif record.POS >=4230 and record.POS <=5093: #5096 minus stop codon #integrase\n",
    "            #print(\"variant\",record.POS,record.POS-4230,record.REF,record.ALT)\n",
    "            alt = record.ALT\n",
    "            temp = []\n",
    "            #temp.append(record.REF)\n",
    "            #temp.append(i_seq[record.POS-4230])\n",
    "            for i in range(len(alt[0])):\n",
    "                base = str(alt[0])[i]\n",
    "                if base not in temp:\n",
    "                    temp.append(base)\n",
    "            i_seq[record.POS-4230] = temp \n",
    "            #print(i_seq[record.POS-4230])\n",
    "            \n",
    "    return protease_seq, rt_seq, i_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [],
   "source": [
    "def createHIVDBRequest(prot,rt,integrase, out):\n",
    "    cmd = 'sierrapy mutations '\n",
    "    #protease\n",
    "    print(\"PROT1 \",len(prot))\n",
    "    for i in prot:\n",
    "        print(\"PROT: \", i)\n",
    "        for j in i[2]:\n",
    "            #print(\"SIERRA: \" , i[0],i[1], j)\n",
    "            mut = \"PR:\"+i[1]+str(i[0])+j\n",
    "            print(\"SIERRA\", mut)\n",
    "            cmd = cmd + \" \"+mut\n",
    "    #RT\n",
    "    for i in rt:\n",
    "        #print(\"PROT: \", i)\n",
    "        for j in i[2]:\n",
    "        #print(\"SIERRA: \" , i[0],i[1], j)\n",
    "            mut = \"RT:\"+i[1]+str(i[0])+j\n",
    "            print(\"SIERRA\", mut)\n",
    "            cmd = cmd + \" \"+mut\n",
    "    #RT\n",
    "    for i in integrase:\n",
    "        #print(\"PROT: \", i)\n",
    "        for j in i[2]:\n",
    "            #print(\"SIERRA: \" , i[0],i[1], j)\n",
    "            mut = \"RT:\"+i[1]+str(i[0])+j\n",
    "            print(\"SIERRA\", mut)\n",
    "            cmd = cmd + \" \"+mut      \n",
    "    #protStr\n",
    "    cmd = cmd + \" -o \"+out\n",
    "    #cmd = 'sierrapy mutations PR:L10I -o output_171.json'\n",
    "    print(cmd)\n",
    "    print (subprocess.check_output(cmd, shell=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "b''\n"
     ]
    }
   ],
   "source": [
    "cmd = \"sierrapy mutations PR:K64V -o output.json\"\n",
    "print (subprocess.check_output(cmd, shell=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MUTATION: 2 Q comb ['H']\n",
      "MUTATION: 3 V comb ['I']\n",
      "MUTATION: 15 I comb ['V']\n",
      "MUTATION: 37 S comb ['D', 'V', 'N', 'I']\n",
      "MUTATION: 41 R comb ['K']\n",
      "MUTATION: 42 W comb ['*']\n",
      "MUTATION: 62 I comb ['V']\n",
      "MUTATION: 64 I comb ['V']\n",
      "MUTATION: 72 I comb ['T']\n",
      "MUTATION: 77 V comb ['I']\n",
      "MUTATION: 92 Q comb ['R', 'P', 'L']\n",
      "MUTATION: 93 I comb ['L']\n",
      "M [[2, 'Q', ['H']], [3, 'V', ['I']], [15, 'I', ['V']], [37, 'S', ['D', 'V', 'N', 'I']], [41, 'R', ['K']], [42, 'W', ['*']], [62, 'I', ['V']], [64, 'I', ['V']], [72, 'I', ['T']], [77, 'V', ['I']], [92, 'Q', ['R', 'P', 'L']], [93, 'I', ['L']]]\n",
      "PROTEASE MUT [[2, 'Q', ['H']], [3, 'V', ['I']], [15, 'I', ['V']], [37, 'S', ['D', 'V', 'N', 'I']], [41, 'R', ['K']], [42, 'W', ['*']], [62, 'I', ['V']], [64, 'I', ['V']], [72, 'I', ['T']], [77, 'V', ['I']], [92, 'Q', ['R', 'P', 'L']], [93, 'I', ['L']]]\n",
      "MUTATION: 6 E comb ['K']\n",
      "MUTATION: 8 V comb ['I']\n",
      "MUTATION: 11 K comb ['R']\n",
      "MUTATION: 26 L comb ['M']\n",
      "MUTATION: 29 E comb ['K']\n",
      "MUTATION: 35 V comb ['I']\n",
      "MUTATION: 44 E comb ['K']\n",
      "MUTATION: 47 I comb ['K', 'N']\n",
      "MUTATION: 48 S comb ['T', 'P']\n",
      "MUTATION: 53 E comb ['K']\n",
      "MUTATION: 60 V comb ['I']\n",
      "MUTATION: 64 K comb ['R']\n",
      "MUTATION: 68 S comb ['G']\n",
      "MUTATION: 76 D comb ['N']\n",
      "M [[6, 'E', ['K']], [8, 'V', ['I']], [11, 'K', ['R']], [26, 'L', ['M']], [29, 'E', ['K']], [35, 'V', ['I']], [44, 'E', ['K']], [47, 'I', ['K', 'N']], [48, 'S', ['T', 'P']], [53, 'E', ['K']], [60, 'V', ['I']], [64, 'K', ['R']], [68, 'S', ['G']], [76, 'D', ['N']]]\n",
      "RT MUT [[6, 'E', ['K']], [8, 'V', ['I']], [11, 'K', ['R']], [26, 'L', ['M']], [29, 'E', ['K']], [35, 'V', ['I']], [44, 'E', ['K']], [47, 'I', ['K', 'N']], [48, 'S', ['T', 'P']], [53, 'E', ['K']], [60, 'V', ['I']], [64, 'K', ['R']], [68, 'S', ['G']], [76, 'D', ['N']]]\n",
      "MUTATION: 10 D comb ['E']\n",
      "MUTATION: 11 E comb ['D']\n",
      "MUTATION: 39 S comb ['G']\n",
      "MUTATION: 50 M comb ['L']\n",
      "MUTATION: 69 E comb ['K']\n",
      "MUTATION: 72 V comb ['I']\n",
      "M [[10, 'D', ['E']], [11, 'E', ['D']], [39, 'S', ['G']], [50, 'M', ['L']], [69, 'E', ['K']], [72, 'V', ['I']]]\n",
      "INT MUT [[10, 'D', ['E']], [11, 'E', ['D']], [39, 'S', ['G']], [50, 'M', ['L']], [69, 'E', ['K']], [72, 'V', ['I']]]\n",
      "PROT1  12\n",
      "PROT:  [2, 'Q', ['H']]\n",
      "SIERRA PR:Q2H\n",
      "PROT:  [3, 'V', ['I']]\n",
      "SIERRA PR:V3I\n",
      "PROT:  [15, 'I', ['V']]\n",
      "SIERRA PR:I15V\n",
      "PROT:  [37, 'S', ['D', 'V', 'N', 'I']]\n",
      "SIERRA PR:S37D\n",
      "SIERRA PR:S37V\n",
      "SIERRA PR:S37N\n",
      "SIERRA PR:S37I\n",
      "PROT:  [41, 'R', ['K']]\n",
      "SIERRA PR:R41K\n",
      "PROT:  [42, 'W', ['*']]\n",
      "SIERRA PR:W42*\n",
      "PROT:  [62, 'I', ['V']]\n",
      "SIERRA PR:I62V\n",
      "PROT:  [64, 'I', ['V']]\n",
      "SIERRA PR:I64V\n",
      "PROT:  [72, 'I', ['T']]\n",
      "SIERRA PR:I72T\n",
      "PROT:  [77, 'V', ['I']]\n",
      "SIERRA PR:V77I\n",
      "PROT:  [92, 'Q', ['R', 'P', 'L']]\n",
      "SIERRA PR:Q92R\n",
      "SIERRA PR:Q92P\n",
      "SIERRA PR:Q92L\n",
      "PROT:  [93, 'I', ['L']]\n",
      "SIERRA PR:I93L\n",
      "SIERRA RT:E6K\n",
      "SIERRA RT:V8I\n",
      "SIERRA RT:K11R\n",
      "SIERRA RT:L26M\n",
      "SIERRA RT:E29K\n",
      "SIERRA RT:V35I\n",
      "SIERRA RT:E44K\n",
      "SIERRA RT:I47K\n",
      "SIERRA RT:I47N\n",
      "SIERRA RT:S48T\n",
      "SIERRA RT:S48P\n",
      "SIERRA RT:E53K\n",
      "SIERRA RT:V60I\n",
      "SIERRA RT:K64R\n",
      "SIERRA RT:S68G\n",
      "SIERRA RT:D76N\n",
      "SIERRA RT:D10E\n",
      "SIERRA RT:E11D\n",
      "SIERRA RT:S39G\n",
      "SIERRA RT:M50L\n",
      "SIERRA RT:E69K\n",
      "SIERRA RT:V72I\n",
      "sierrapy mutations  PR:Q2H PR:V3I PR:I15V PR:S37D PR:S37V PR:S37N PR:S37I PR:R41K PR:W42* PR:I62V PR:I64V PR:I72T PR:V77I PR:Q92R PR:Q92P PR:Q92L PR:I93L RT:E6K RT:V8I RT:K11R RT:L26M RT:E29K RT:V35I RT:E44K RT:I47K RT:I47N RT:S48T RT:S48P RT:E53K RT:V60I RT:K64R RT:S68G RT:D76N RT:D10E RT:E11D RT:S39G RT:M50L RT:E69K RT:V72I -o output_171.json\n",
      "b''\n"
     ]
    }
   ],
   "source": [
    "vcfFile = '/Users/jenniferureta/Desktop/bioinfo/HIV/sequence/HIV2-171/171_mult.vcf'\n",
    "prot,rt,i = insertVariants_mult(vcfFile,\"protease.txt\",\"reverse_transcriptase.txt\",\"integrase.txt\")\n",
    "refp = \"protease_amino.txt\"\n",
    "refrt = \"rt_amino.txt\"\n",
    "refi = \"integrase_amino.txt\"\n",
    "protease_mut = getMutations_mult(prot,refp)\n",
    "print(\"PROTEASE MUT\",protease_mut)\n",
    "rt_mut = getMutations_mult(rt,refrt)\n",
    "print(\"RT MUT\",rt_mut)\n",
    "int_mut = getMutations_mult(i,refi)\n",
    "print(\"INT MUT\",int_mut)\n",
    "createHIVDBRequest(protease_mut, rt_mut,int_mut,\"output_171.json\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [],
   "source": [
    "def insertVariants(vcfFile,pFileName, rtFileName, iFileName): \n",
    "    #one mutation per position and returns amino acids\n",
    "    vcf_reader = vcf.Reader(open(vcfFile,'r'))\n",
    "    \n",
    "    pFile = open(pFileName, 'r')\n",
    "    protease_seq = [ch for ch in pFile.read()]\n",
    "    #print(\"PROT\", protease_seq)\n",
    "    rtFile = open(rtFileName, 'r')\n",
    "    rt_seq = [ch for ch in rtFile.read()]   \n",
    "    #print(\"RT\", rt_seq)\n",
    "    iFile = open(iFileName, 'r')\n",
    "    i_seq = [ch for ch in iFile.read()]\n",
    "    #print(\"I\", i_seq)\n",
    "    #lofreq starts at 1 but our indexing starts at 0 so always subtract 1 from the position given\n",
    "    \n",
    "    for record in vcf_reader:\n",
    "       \n",
    "        if record.POS >=2253 and record.POS <2550: #protease\n",
    "            print(\"variant\",record.POS,record.POS-2253,record.REF,record.ALT)\n",
    "            alt = record.ALT[0]\n",
    "            protease_seq[record.POS-2253] = alt\n",
    "        elif record.POS >=2550 and record.POS < 4229: #3869: #reverse transcriptase\n",
    "            print(\"variant\",record.POS,record.POS-2550,record.REF,record.ALT)\n",
    "            alt = record.ALT[0]\n",
    "            rt_seq[record.POS-2550] = alt\n",
    "        elif record.POS >=4230 and record.POS <=5093: #5096 minus stop codon #integrase\n",
    "            print(\"variant\",record.POS,record.POS-4230,record.REF,record.ALT)\n",
    "            alt = record.ALT[0]\n",
    "            i_seq[record.POS-4230] = alt\n",
    "            \n",
    "   # print(\"PROT VAR\\n\",protease_seq) \n",
    "    #print(\"RT VAR\\n\",rt_seq) \n",
    "    #print(\"I VAR\\n\",i_seq) \n",
    "    return protease_seq, rt_seq, i_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getMutations(pseq, rtseq, iseq, refP, refRT, refI):\n",
    "    #protease\n",
    "    \n",
    "    pmutations=[]\n",
    "    rtmutations=[]\n",
    "    imutations=[]\n",
    "    plen = len(pseq)\n",
    "    ptranslation = \"\"\n",
    "    for i in range(0,plen,3): #translation to amino acid\n",
    "        codonStr = str(pseq[i])+str(pseq[i+1])+str(pseq[i+2])\n",
    "        #print(\"CODON\",pseq[i],pseq[i+1],pseq[i+2], codonStr)\n",
    "        ptranslation = ptranslation + aminoAcids.get(codonStr,\"!\")\n",
    "    #output file format POSITION REF ALT\n",
    "    pf = open(refP, 'r')\n",
    "    ref_protease = [ch for ch in pf.read()]\n",
    "    plen = len(ptranslation)\n",
    "\n",
    "    for i in range(0,plen,1):\n",
    "        if ref_protease[i] != ptranslation[i]:\n",
    "            print(\"M\", i+1,ref_protease[i],  ptranslation[i])\n",
    "            pmutations.append([i+1,ref_protease[i],ptranslation[i]])\n",
    "    #reverse transcriptase\n",
    "    rtlen = len(rtseq)\n",
    "    rttranslation = \"\"\n",
    "    for i in range(0,rtlen,3):\n",
    "        codonStr = str(rtseq[i])+str(rtseq[i+1])+str(rtseq[i+2])\n",
    "        #print(\"CODON\",rtseq[i],rtseq[i+1],rtseq[i+2], codonStr)\n",
    "        rttranslation = rttranslation + aminoAcids.get(codonStr,\"!\")\n",
    "       \n",
    "    #output file format POSITION REF ALT\n",
    "    rf = open(refRT, 'r')\n",
    "    ref_RT = [ch for ch in rf.read()]\n",
    "    rtlen = len(rttranslation)\n",
    "\n",
    "    for i in range(0,rtlen,1):\n",
    "        if ref_RT[i] != rttranslation[i]:\n",
    "            print(\"M\", i+1,ref_RT[i],  rttranslation[i])\n",
    "            rtmutations.append([i+1,ref_RT[i],rttranslation[i]])\n",
    "    #integrase\n",
    "    ilen = len(iseq)\n",
    "    itranslation = \"\"\n",
    "    for i in range(0,ilen,3):\n",
    "        codonStr = str(iseq[i])+str(iseq[i+1])+str(iseq[i+2])\n",
    "        #print(\"CODON\",iseq[i],iseq[i+1],iseq[i+2], codonStr)\n",
    "        itranslation = itranslation + aminoAcids.get(codonStr,\"!\")\n",
    "    #output file format POSITION REF ALT\n",
    "    intf = open(refI, 'r')\n",
    "    ref_int = [ch for ch in intf.read()]\n",
    "    ilen = len(itranslation)\n",
    "\n",
    "    for i in range(0,ilen,1):\n",
    "        if ref_int[i] != itranslation[i]:\n",
    "            print(\"M\", i+1,ref_int[i],  itranslation[i])\n",
    "            imutations.append([i+1,ref_int[i],itranslation[i]])\n",
    "    print(pmutations)\n",
    "    print(rtmutations)\n",
    "    print(imutations)\n",
    "    return pmutations,rtmutations,imutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Get Sequence\n",
      "Get Sequence\n",
      "Get Sequence\n",
      "length of sequence 297\n",
      "translated PQVTLWQRPLVTIKIGGQLKEALLDTGADDTVLEEMSLPGRWKPKMIGGIGGFIKVRQYDQILIEICGHKAIGTVLVGPTPVNIIGRNLLTQIGCTLNF\n",
      "length of sequence 1680\n",
      "translated PISPIETVPVKLKPGMDGPKVKQWPLTEEKIKALVEICTEMEKEGKISKIGPENPYNTPVFAIKKKDSTKWRKLVDFRELNKRTQDFWEVQLGIPHPAGLKKKKSVTVLDVGDAYFSVPLDEDFRKYTAFTIPSINNETPGIRYQYNVLPQGWKGSPAIFQSSMTKILEPFRKQNPDIVIYQYMDDLYVGSDLEIGQHRTKIEELRQHLLRWGLTTPDKKHQKEPPFLWMGYELHPDKWTVQPIVLPEKDSWTVNDIQKLVGKLNWASQIYPGIKVRQLCKLLRGTKALTEVIPLTEEAELELAENREILKEPVHGVYYDPSKDLIAEIQKQGQGQWTYQIYQEPFKNLKTGKYARMRGAHTNDVKQLTEAVQKITTESIVIWGKTPKFKLPIQKETWETWWTEYWQATWIPEWEFVNTPPLVKLWYQLEKEPIVGAETFYVDGAANRETKLGKAGYVTNRGRQKVVTLTDTTNQKTELQAIYLALQDSGLEVNIVTDSQYALGIIQAQPDQSESELVNQIIEQLIKKEKVYLAWVPAHKGIGGNEQVDKLVSAGIRKVL\n",
      "length of sequence 864\n",
      "translated FLDGIDKAQDEHEKYHSNWRAMASDFNLPPVVAKEIVASCDKCQLKGEAMHGQVDCSPGIWQLDCTHLEGKVILVAVHVASGYIEAEVIPAETGQETAYFLLKLAGRWPVKTIHTDNGSNFTGATVRAACWWAGIKQEFGIPYNPQSQGVVESMNKELKKIIGQVRDQAEHLKTAVQMAVFIHNFKRKGGIGGYSAGERIVDIIATDIQTKELQKQITKIQNFRVYYRDSRNPLWKGPAKLLWKGEGAVVIQDNSDIKVVPRRKAKIIRDYGKQMAGDDCVASRQDED\n",
      "variant 2258 5 G [A]\n",
      "variant 2259 6 G [A]\n",
      "variant 2282 29 C [T]\n",
      "variant 2294 41 G [A]\n",
      "variant 2295 42 A [G]\n",
      "variant 2303 50 G [A]\n",
      "variant 2306 53 A [G]\n",
      "variant 2312 59 G [A]\n",
      "variant 2321 68 A [C]\n",
      "variant 2361 108 A [G]\n",
      "variant 2362 109 G [A]\n",
      "variant 2366 113 G [A]\n",
      "variant 2372 119 A [C]\n",
      "variant 2374 121 G [A]\n",
      "variant 2378 125 G [A]\n",
      "variant 2426 173 G [A]\n",
      "variant 2436 183 A [G]\n",
      "variant 2442 189 A [G]\n",
      "variant 2447 194 A [G]\n",
      "variant 2453 200 T [C]\n",
      "variant 2456 203 A [G]\n",
      "variant 2462 209 A [G]\n",
      "variant 2467 214 T [C]\n",
      "variant 2468 215 A [G]\n",
      "variant 2477 224 A [G]\n",
      "variant 2481 228 G [A]\n",
      "variant 2486 233 A [G]\n",
      "variant 2489 236 T [C]\n",
      "variant 2495 242 T [C]\n",
      "variant 2516 263 T [C]\n",
      "variant 2527 274 A [G]\n",
      "variant 2529 276 A [C]\n",
      "variant 2537 284 C [T]\n",
      "variant 2558 8 C [T]\n",
      "variant 2565 15 G [A]\n",
      "variant 2567 17 G [A]\n",
      "variant 2570 20 T [G]\n",
      "variant 2571 21 G [A]\n",
      "variant 2581 31 A [G]\n",
      "variant 2603 53 C [T]\n",
      "variant 2625 75 T [A]\n",
      "variant 2633 83 A [G]\n",
      "variant 2634 84 G [A]\n",
      "variant 2652 102 G [A]\n",
      "variant 2657 107 A [G]\n",
      "variant 2663 113 T [C]\n",
      "variant 2669 119 G [A]\n",
      "variant 2679 129 G [A]\n",
      "variant 2684 134 G [A]\n",
      "variant 2689 139 T [A]\n",
      "variant 2690 140 T [A]\n",
      "variant 2691 141 T [A]\n",
      "variant 2706 156 G [A]\n",
      "variant 2717 167 C [T]\n",
      "variant 2723 173 T [C]\n",
      "variant 2727 177 G [A]\n",
      "variant 2740 190 A [G]\n",
      "variant 2751 201 A [G]\n",
      "variant 2753 203 T [C]\n",
      "variant 2775 225 G [A]\n",
      "variant 2777 227 T [C]\n",
      "variant 2789 239 T [A]\n",
      "variant 2795 245 G [A]\n",
      "variant 2798 248 A [G]\n",
      "variant 2807 257 C [T]\n",
      "variant 2840 290 C [T]\n",
      "variant 2846 296 G [A]\n",
      "variant 2853 303 A [C]\n",
      "variant 2857 307 A [G]\n",
      "variant 2860 310 A [G]\n",
      "variant 2879 329 T [C]\n",
      "variant 2885 335 T [C]\n",
      "variant 2894 344 T [C]\n",
      "variant 2903 353 T [G]\n",
      "variant 2913 363 G [A]\n",
      "variant 2916 366 G [A]\n",
      "variant 2918 368 C [T]\n",
      "variant 2924 374 G [A]\n",
      "variant 2927 377 G [A]\n",
      "variant 2939 389 T [C]\n",
      "variant 2942 392 C [T]\n",
      "variant 2972 422 G [A]\n",
      "variant 2990 440 T [C]\n",
      "variant 2999 449 A [G]\n",
      "variant 3002 452 G [A]\n",
      "variant 3008 458 G [A]\n",
      "variant 3017 467 A [G]\n",
      "variant 3034 484 G [A]\n",
      "variant 3034 484 G [C]\n",
      "variant 3046 496 A [G]\n",
      "variant 3047 497 A [G]\n",
      "variant 3051 501 T [C]\n",
      "variant 3056 506 G [A]\n",
      "variant 3062 512 T [C]\n",
      "variant 3066 516 A [G]\n",
      "variant 3068 518 A [G]\n",
      "variant 3080 530 C [T]\n",
      "variant 3084 534 G [A]\n",
      "variant 3086 536 T [C]\n",
      "variant 3089 539 C [T]\n",
      "variant 3092 542 T [C]\n",
      "variant 3105 555 G [A]\n",
      "variant 3110 560 G [A]\n",
      "variant 3118 568 G [A]\n",
      "variant 3122 572 T [C]\n",
      "variant 3125 575 C [T]\n",
      "variant 3129 579 G [A]\n",
      "variant 3136 586 G [A]\n",
      "variant 3137 587 G [A]\n",
      "variant 3138 588 C [A]\n",
      "variant 3140 590 G [A]\n",
      "variant 3143 593 T [C]\n",
      "variant 3148 598 C [T]\n",
      "variant 3153 603 A [G]\n",
      "variant 3161 611 G [A]\n",
      "variant 3167 617 A [G]\n",
      "variant 3180 630 A [G]\n",
      "variant 3182 632 G [A]\n",
      "variant 3189 639 C [T]\n",
      "variant 3191 641 T [C]\n",
      "variant 3197 647 A [G]\n",
      "variant 3206 656 A [G]\n",
      "variant 3209 659 A [G]\n",
      "variant 3215 665 G [A]\n",
      "variant 3218 668 A [G]\n",
      "variant 3272 722 A [G]\n",
      "variant 3279 729 A [G]\n",
      "variant 3284 734 G [A]\n",
      "variant 3285 735 C [T]\n",
      "variant 3287 737 G [A]\n",
      "variant 3293 743 A [G]\n",
      "variant 3296 746 A [G]\n",
      "variant 3301 751 G [A]\n",
      "variant 3317 767 C [T]\n",
      "variant 3323 773 G [A]\n",
      "variant 3335 785 G [A]\n",
      "variant 3339 789 T [C]\n",
      "variant 3341 791 G [A]\n",
      "variant 3344 794 T [C]\n",
      "variant 3348 798 G [A]\n",
      "variant 3350 800 A [G]\n",
      "variant 3362 812 C [T]\n",
      "variant 3363 813 C [T]\n",
      "variant 3368 818 G [A]\n",
      "variant 3371 821 T [C]\n",
      "variant 3375 825 G [A]\n",
      "variant 3379 829 G [A]\n",
      "variant 3395 845 C [T]\n",
      "variant 3400 850 G [A]\n",
      "variant 3401 851 A [G]\n",
      "variant 3426 876 A [G]\n",
      "variant 3437 887 A [G]\n",
      "variant 3458 908 G [A]\n",
      "variant 3470 920 A [G]\n",
      "variant 3473 923 G [A]\n",
      "variant 3483 933 G [A]\n",
      "variant 3484 934 A [G]\n",
      "variant 3485 935 A [G]\n",
      "variant 3497 947 A [G]\n",
      "variant 3500 950 G [A]\n",
      "variant 3514 964 C [T]\n",
      "variant 3525 975 A [G]\n",
      "variant 3548 998 G [A]\n",
      "variant 3557 1007 A [G]\n",
      "variant 3608 1058 A [G]\n",
      "variant 3616 1066 G [A]\n",
      "variant 3618 1068 A [G]\n",
      "variant 3642 1092 G [A]\n",
      "variant 3647 1097 A [G]\n",
      "variant 3650 1100 A [G]\n",
      "variant 3659 1109 G [A]\n",
      "variant 3665 1115 G [A]\n",
      "variant 3668 1118 A [G]\n",
      "variant 3675 1125 A [G]\n",
      "variant 3690 1140 G [A]\n",
      "variant 3722 1172 G [A]\n",
      "variant 3725 1175 C [T]\n",
      "variant 3734 1184 G [A]\n",
      "variant 3737 1187 A [G]\n",
      "variant 3746 1196 A [G]\n",
      "variant 3747 1197 A [G]\n",
      "variant 3757 1207 C [T]\n",
      "variant 3758 1208 A [G]\n",
      "variant 3800 1250 T [C]\n",
      "variant 3818 1268 G [A]\n",
      "variant 3833 1283 G [A]\n",
      "variant 3837 1287 G [A]\n",
      "variant 3839 1289 G [A]\n",
      "variant 3878 1328 T [C]\n",
      "variant 3890 1340 C [T]\n",
      "variant 3893 1343 G [A]\n",
      "variant 3896 1346 G [A]\n",
      "variant 3903 1353 T [C]\n",
      "variant 3908 1358 A [G]\n",
      "variant 3923 1373 T [C]\n",
      "variant 3927 1377 A [G]\n",
      "variant 3931 1381 G [A]\n",
      "variant 3932 1382 A [G]\n",
      "variant 3944 1394 A [G]\n",
      "variant 3947 1397 T [C]\n",
      "variant 3951 1401 A [T]\n",
      "variant 3965 1415 A [C]\n",
      "variant 3968 1418 A [G]\n",
      "variant 3977 1427 G [A]\n",
      "variant 3984 1434 T [C]\n",
      "variant 3989 1439 A [G]\n",
      "variant 3996 1446 T [C]\n",
      "variant 3999 1449 C [T]\n",
      "variant 4007 1457 G [A]\n",
      "variant 4010 1460 G [A]\n",
      "variant 4016 1466 G [A]\n",
      "variant 4025 1475 A [G]\n",
      "variant 4028 1478 A [G]\n",
      "variant 4031 1481 C [T]\n",
      "variant 4037 1487 A [G]\n",
      "variant 4043 1493 C [T]\n",
      "variant 4046 1496 A [G]\n",
      "variant 4046 1496 A [T]\n",
      "variant 4056 1506 T [C]\n",
      "variant 4083 1533 C [A]\n",
      "variant 4085 1535 A [G]\n",
      "variant 4091 1541 A [G]\n",
      "variant 4097 1547 G [A]\n",
      "variant 4105 1555 A [C]\n",
      "variant 4105 1555 A [G]\n",
      "variant 4106 1556 T [C]\n",
      "variant 4106 1556 T [G]\n",
      "variant 4121 1571 G [A]\n",
      "variant 4134 1584 G [A]\n",
      "variant 4139 1589 G [A]\n",
      "variant 4145 1595 T [C]\n",
      "variant 4148 1598 G [A]\n",
      "variant 4166 1616 C [T]\n",
      "variant 4169 1619 A [G]\n",
      "variant 4181 1631 A [G]\n",
      "variant 4187 1637 A [G]\n",
      "variant 4189 1639 A [G]\n",
      "variant 4190 1640 A [G]\n",
      "variant 4196 1646 T [C]\n",
      "variant 4205 1655 C [T]\n",
      "variant 4209 1659 G [A]\n",
      "variant 4211 1661 T [A]\n",
      "variant 4217 1667 C [T]\n",
      "variant 4222 1672 A [G]\n",
      "variant 4223 1673 A [G]\n",
      "variant 4235 5 A [G]\n",
      "variant 4256 26 A [G]\n",
      "variant 4259 29 T [A]\n",
      "variant 4262 32 A [C]\n",
      "variant 4268 38 G [A]\n",
      "variant 4271 41 A [G]\n",
      "variant 4313 83 G [A]\n",
      "variant 4334 104 A [G]\n",
      "variant 4344 114 A [G]\n",
      "variant 4349 119 T [C]\n",
      "variant 4352 122 T [C]\n",
      "variant 4367 137 A [G]\n",
      "variant 4370 140 A [G]\n",
      "variant 4377 147 A [C]\n",
      "variant 4382 152 T [C]\n",
      "variant 4385 155 A [G]\n",
      "variant 4397 167 T [C]\n",
      "variant 4416 186 C [T]\n",
      "variant 4424 194 T [C]\n",
      "variant 4433 203 A [G]\n",
      "variant 4434 204 G [A]\n",
      "variant 4439 209 A [G]\n",
      "variant 4443 213 G [A]\n",
      "variant 4466 236 A [G]\n",
      "variant 4469 239 C [T]\n",
      "variant 4484 254 A [G]\n",
      "variant 4508 278 A [C]\n",
      "variant 4526 296 T [C]\n",
      "variant 4550 320 A [G]\n",
      "variant 4561 331 A [G]\n",
      "variant 4564 334 C [T]\n",
      "variant 4568 338 A [C]\n",
      "variant 4574 344 T [A]\n",
      "variant 4584 354 A [C]\n",
      "variant 4585 355 G [C]\n",
      "variant 4586 356 C [T]\n",
      "variant 4596 366 G [A]\n",
      "variant 4599 369 G [A]\n",
      "variant 4604 374 G [A]\n",
      "variant 4609 379 G [A]\n",
      "variant 4613 383 C [T]\n",
      "variant 4628 398 G [A]\n",
      "variant 4631 401 A [G]\n",
      "variant 4634 404 C [T]\n",
      "variant 4637 407 G [A]\n",
      "variant 4640 410 G [A]\n",
      "variant 4641 411 G [A]\n",
      "variant 4643 413 A [G]\n",
      "variant 4649 419 A [C]\n",
      "variant 4683 453 G [A]\n",
      "variant 4691 461 G [A]\n",
      "variant 4721 491 G [A]\n",
      "variant 4736 506 T [C]\n",
      "variant 4742 512 T [C]\n",
      "variant 4813 583 G [C]\n",
      "variant 4826 596 A [G]\n",
      "variant 4866 636 T [C]\n",
      "variant 4895 665 T [C]\n",
      "variant 4913 683 G [A]\n",
      "variant 4918 688 G [A]\n",
      "variant 4923 693 A [G]\n",
      "variant 4928 698 A [G]\n",
      "variant 4931 701 T [C]\n",
      "variant 4949 719 G [A]\n",
      "variant 4952 722 C [T]\n",
      "variant 4970 740 G [A]\n",
      "variant 4988 758 T [C]\n",
      "variant 4991 761 T [C]\n",
      "variant 4994 764 T [C]\n",
      "variant 5006 776 A [G]\n",
      "variant 5009 779 G [A]\n",
      "variant 5027 797 G [A]\n",
      "variant 5033 803 T [C]\n",
      "variant 5076 846 A [G]\n",
      "M 3 V I\n",
      "M 15 I V\n",
      "M 37 S D\n",
      "M 41 R K\n",
      "M 42 W *\n",
      "M 62 I V\n",
      "M 64 I V\n",
      "M 72 I T\n",
      "M 77 V I\n",
      "M 92 Q R\n",
      "M 93 I L\n",
      "M 6 E K\n",
      "M 8 V I\n",
      "M 11 K R\n",
      "M 26 L M\n",
      "M 29 E K\n",
      "M 35 V I\n",
      "M 44 E K\n",
      "M 47 I K\n",
      "M 48 S T\n",
      "M 53 E K\n",
      "M 60 V I\n",
      "M 64 K R\n",
      "M 68 S G\n",
      "M 76 D N\n",
      "M 102 K Q\n",
      "M 103 K R\n",
      "M 104 K R\n",
      "M 122 E K\n",
      "M 123 D N\n",
      "M 153 W *\n",
      "M 162 S T\n",
      "M 166 K R\n",
      "M 173 K E\n",
      "M 179 V I\n",
      "M 186 D N\n",
      "M 190 G E\n",
      "M 194 E K\n",
      "M 196 G E\n",
      "M 197 Q K\n",
      "M 200 T I\n",
      "M 202 I V\n",
      "M 211 R G\n",
      "M 214 L F\n",
      "M 244 I V\n",
      "M 251 S N\n",
      "M 267 A T\n",
      "M 272 P S\n",
      "M 276 V I\n",
      "M 277 R K\n",
      "M 284 R K\n",
      "M 293 I V\n",
      "M 312 E R\n",
      "M 322 S L\n",
      "M 326 I V\n",
      "M 356 R K\n",
      "M 357 M V\n",
      "M 365 V I\n",
      "M 376 T A\n",
      "M 381 V I\n",
      "M 400 T A\n",
      "M 403 T M\n",
      "M 430 E K\n",
      "M 460 N D\n",
      "M 461 R K\n",
      "M 468 T S\n",
      "M 483 Y H\n",
      "M 512 Q K\n",
      "M 519 N R\n",
      "M 529 E K\n",
      "M 547 Q R\n",
      "M 554 A T\n",
      "M 558 K R\n",
      "M 10 D E\n",
      "M 11 E D\n",
      "M 39 S G\n",
      "M 50 M L\n",
      "M 69 E K\n",
      "M 72 V I\n",
      "M 111 K R\n",
      "M 112 T I\n",
      "M 119 S P\n",
      "M 123 G S\n",
      "M 124 A T\n",
      "M 127 R K\n",
      "M 138 E K\n",
      "M 152 E K\n",
      "M 154 M I\n",
      "M 195 S T\n",
      "M 230 S N\n",
      "M 232 N D\n",
      "M 283 S G\n",
      "[[3, 'V', 'I'], [15, 'I', 'V'], [37, 'S', 'D'], [41, 'R', 'K'], [42, 'W', '*'], [62, 'I', 'V'], [64, 'I', 'V'], [72, 'I', 'T'], [77, 'V', 'I'], [92, 'Q', 'R'], [93, 'I', 'L']]\n",
      "[[6, 'E', 'K'], [8, 'V', 'I'], [11, 'K', 'R'], [26, 'L', 'M'], [29, 'E', 'K'], [35, 'V', 'I'], [44, 'E', 'K'], [47, 'I', 'K'], [48, 'S', 'T'], [53, 'E', 'K'], [60, 'V', 'I'], [64, 'K', 'R'], [68, 'S', 'G'], [76, 'D', 'N'], [102, 'K', 'Q'], [103, 'K', 'R'], [104, 'K', 'R'], [122, 'E', 'K'], [123, 'D', 'N'], [153, 'W', '*'], [162, 'S', 'T'], [166, 'K', 'R'], [173, 'K', 'E'], [179, 'V', 'I'], [186, 'D', 'N'], [190, 'G', 'E'], [194, 'E', 'K'], [196, 'G', 'E'], [197, 'Q', 'K'], [200, 'T', 'I'], [202, 'I', 'V'], [211, 'R', 'G'], [214, 'L', 'F'], [244, 'I', 'V'], [251, 'S', 'N'], [267, 'A', 'T'], [272, 'P', 'S'], [276, 'V', 'I'], [277, 'R', 'K'], [284, 'R', 'K'], [293, 'I', 'V'], [312, 'E', 'R'], [322, 'S', 'L'], [326, 'I', 'V'], [356, 'R', 'K'], [357, 'M', 'V'], [365, 'V', 'I'], [376, 'T', 'A'], [381, 'V', 'I'], [400, 'T', 'A'], [403, 'T', 'M'], [430, 'E', 'K'], [460, 'N', 'D'], [461, 'R', 'K'], [468, 'T', 'S'], [483, 'Y', 'H'], [512, 'Q', 'K'], [519, 'N', 'R'], [529, 'E', 'K'], [547, 'Q', 'R'], [554, 'A', 'T'], [558, 'K', 'R']]\n",
      "[[10, 'D', 'E'], [11, 'E', 'D'], [39, 'S', 'G'], [50, 'M', 'L'], [69, 'E', 'K'], [72, 'V', 'I'], [111, 'K', 'R'], [112, 'T', 'I'], [119, 'S', 'P'], [123, 'G', 'S'], [124, 'A', 'T'], [127, 'R', 'K'], [138, 'E', 'K'], [152, 'E', 'K'], [154, 'M', 'I'], [195, 'S', 'T'], [230, 'S', 'N'], [232, 'N', 'D'], [283, 'S', 'G']]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "\"\\ncodon = []\\n\\nproteaseStart = 2252 #original protease start is 2253 for index starting at 1\\nrtStart = 2549 #original 2550\\nintegrasetStart = 4229 #original 4230\\nvcf_reader = vcf.Reader(open('/Users/jenniferureta/Desktop/bioinfo/HIV/sequence/HIV2-171/171_sub.vcf','r'))\\n#lofreq starts at 1 but our indexing starts at 0 so always subtract 1 from the position given\\nmutations=[]\\nfor record in vcf_reader:\\n    if record.POS >=2253 or record.POS <2550: #for protease gene 2253-2550\\n        mutation_position = int(record.POS)\\n        pCodon = getCodon(proteaseStart,mutation_position-1,record.ALT) \\n        codonStr = pCodon[0]+pCodon[1]+pCodon[2]\\n        print(codonStr)\\n        print(aminoAcids.get(codonStr,0)) #0 if not found in amino acid table\\n#for reverse transcriptase gene 2550-3869\\n\\n\\n#for integrase gene 4230-5096\\n\\n    print(record)\\n    print (record.POS)\\n\""
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "getGeneRegions()\n",
    "getAminoAcids(\"protease.txt\", \"protease_amino.txt\")\n",
    "getAminoAcids(\"reverse_transcriptase.txt\", \"rt_amino.txt\")\n",
    "getAminoAcids(\"integrase.txt\", \"integrase_amino.txt\")\n",
    "vcfFile = '/Users/jenniferureta/Desktop/bioinfo/HIV/sequence/HIV2-171/171_sub.vcf'\n",
    "pseq,rtseq,iseq = insertVariants(vcfFile,\"protease.txt\",\"reverse_transcriptase.txt\",\"integrase.txt\")\n",
    "pm,rm,im= getMutations(pseq,rtseq,iseq,\"protease_amino.txt\", \"rt_amino.txt\", \"integrase_amino.txt\")\n",
    "#connectToHIVDB()\n",
    "\"\"\"\n",
    "codon = []\n",
    "\n",
    "proteaseStart = 2252 #original protease start is 2253 for index starting at 1\n",
    "rtStart = 2549 #original 2550\n",
    "integrasetStart = 4229 #original 4230\n",
    "vcf_reader = vcf.Reader(open('/Users/jenniferureta/Desktop/bioinfo/HIV/sequence/HIV2-171/171_sub.vcf','r'))\n",
    "#lofreq starts at 1 but our indexing starts at 0 so always subtract 1 from the position given\n",
    "mutations=[]\n",
    "for record in vcf_reader:\n",
    "    if record.POS >=2253 or record.POS <2550: #for protease gene 2253-2550\n",
    "        mutation_position = int(record.POS)\n",
    "        pCodon = getCodon(proteaseStart,mutation_position-1,record.ALT) \n",
    "        codonStr = pCodon[0]+pCodon[1]+pCodon[2]\n",
    "        print(codonStr)\n",
    "        print(aminoAcids.get(codonStr,0)) #0 if not found in amino acid table\n",
    "#for reverse transcriptase gene 2550-3869\n",
    "\n",
    "\n",
    "#for integrase gene 4230-5096\n",
    "\n",
    "    print(record)\n",
    "    print (record.POS)\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convertSAMtoBAM(input_file, output_file):\n",
    "    command = \"samtools view -S -b \"+input_file+\" > \"+output_file\n",
    "    print(subprocess.call(command, shell = True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sortBAM(input_file, output_file):\n",
    "    command = \"samtools sort -o \"+output_file+input_file\n",
    "    print(subprocess.call(command, shell = True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def indexBAM(file)\n",
    "    command = \"samtools -b \"+ input_file\n",
    "    print(subprocess.call(command, shell = True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def performAlignment(in_dir, out_dir, log_file, index):\n",
    "    for entry in os.scandir(in_dir):\n",
    "        if entry.is_dir() and entry.name.startswith('HIV2'):\n",
    "            #print(entry.name)\n",
    "            file1 = in_dir+\"/\"+entry.name+\"/\"+entry.name +\"_1.fastq\"\n",
    "            #print(file1)\n",
    "            file2 = in_dir+\"/\"+entry.name+\"/\"+entry.name +\"_2.fastq\"\n",
    "            #print(file2)\n",
    "            out = out_dir+\"/\"+entry.name+\".sam\"\n",
    "            print(\"OUT\",out)\n",
    "            #print(log_file)\n",
    "            print(\"Path\",os.path.exists(out_dir))\n",
    "            if not os.path.exists(out_dir):\n",
    "             #   print(\"here\")\n",
    "                os.mkdir(out_dir,0o777)                \n",
    "            command = \"(bowtie2 --local -q -x \"+ index + \" -1 \"+file1+ \" -2 \"+ file2 +\" -S \"+ out+\" )2>> \"+log_file\n",
    "            print(command)\n",
    "            processing = \"(echo -e \"+\"\\\"\\\\n\\\\n\"+entry.name+ \"\\\\n\\\\n\\\")>>\"+log_file \n",
    "            print(\"Processing \", processing)\n",
    "            subprocess.call(processing, shell= True)\n",
    "            print(subprocess.call(command,shell=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OUT /Users/jenniferureta/Desktop/bioinfo/HIV/sequence_out/HIV2-157.sam\n",
      "Path True\n",
      "(bowtie2 --local -q -x /Users/jenniferureta/Desktop/bioinfo/hxb2 -1 /Users/jenniferureta/Desktop/bioinfo/HIV/sequence/HIV2-157/HIV2-157_1.fastq -2 /Users/jenniferureta/Desktop/bioinfo/HIV/sequence/HIV2-157/HIV2-157_2.fastq -S /Users/jenniferureta/Desktop/bioinfo/HIV/sequence_out/HIV2-157.sam )2>> /Users/jenniferureta/Desktop/bioinfo/HIV/sequence_out/file_test.log\n",
      "Processing  (echo -e \"\\n\\nHIV2-157\\n\\n\")>>/Users/jenniferureta/Desktop/bioinfo/HIV/sequence_out/file_test.log\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-5-a31272101661>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mlogf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"/Users/jenniferureta/Desktop/bioinfo/HIV/sequence_out/file_test.log\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mindex\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"/Users/jenniferureta/Desktop/bioinfo/hxb2\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m \u001b[0mperformAlignment\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindir\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0moutdir\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mlogf\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-2-0b7cff3f9801>\u001b[0m in \u001b[0;36mperformAlignment\u001b[0;34m(in_dir, out_dir, log_file, index)\u001b[0m\n\u001b[1;32m     19\u001b[0m             \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Processing \"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mprocessing\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     20\u001b[0m             \u001b[0msubprocess\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcall\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprocessing\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mshell\u001b[0m\u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 21\u001b[0;31m             \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msubprocess\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcall\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcommand\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mshell\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/anaconda3/lib/python3.7/subprocess.py\u001b[0m in \u001b[0;36mcall\u001b[0;34m(timeout, *popenargs, **kwargs)\u001b[0m\n\u001b[1;32m    317\u001b[0m     \u001b[0;32mwith\u001b[0m \u001b[0mPopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0mpopenargs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mp\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    318\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 319\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    320\u001b[0m         \u001b[0;32mexcept\u001b[0m\u001b[0;34m:\u001b[0m  \u001b[0;31m# Including KeyboardInterrupt, wait handled that.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    321\u001b[0m             \u001b[0mp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkill\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/subprocess.py\u001b[0m in \u001b[0;36mwait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    982\u001b[0m             \u001b[0mendtime\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_time\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    983\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 984\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_wait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    985\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mKeyboardInterrupt\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    986\u001b[0m             \u001b[0;31m# https://bugs.python.org/issue25942\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/subprocess.py\u001b[0m in \u001b[0;36m_wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m   1616\u001b[0m                         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreturncode\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1617\u001b[0m                             \u001b[0;32mbreak\u001b[0m  \u001b[0;31m# Another thread waited.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1618\u001b[0;31m                         \u001b[0;34m(\u001b[0m\u001b[0mpid\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msts\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_try_wait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1619\u001b[0m                         \u001b[0;31m# Check the pid and loop as waitpid has been known to\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1620\u001b[0m                         \u001b[0;31m# return 0 even without WNOHANG in odd situations.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/subprocess.py\u001b[0m in \u001b[0;36m_try_wait\u001b[0;34m(self, wait_flags)\u001b[0m\n\u001b[1;32m   1574\u001b[0m             \u001b[0;34m\"\"\"All callers to this function MUST hold self._waitpid_lock.\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1575\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1576\u001b[0;31m                 \u001b[0;34m(\u001b[0m\u001b[0mpid\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msts\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwaitpid\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpid\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mwait_flags\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1577\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mChildProcessError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1578\u001b[0m                 \u001b[0;31m# This happens if SIGCLD is set to be ignored or waiting\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "indir = \"/Users/jenniferureta/Desktop/bioinfo/HIV/sequence\"\n",
    "outdir = \"/Users/jenniferureta/Desktop/bioinfo/HIV/sequence_out\"\n",
    "logf = \"/Users/jenniferureta/Desktop/bioinfo/HIV/sequence_out/file_test.log\"\n",
    "index = \"/Users/jenniferureta/Desktop/bioinfo/hxb2\"\n",
    "performAlignment(indir,outdir,logf,index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'convertSAMtoBAM' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-4-4fb427f5a7ef>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0minputf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"/Users/jenniferureta/Desktop/bioinfo/RESULTS_LOCAL/HIV2_171.sam \"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0moutputf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"/Users/jenniferureta/Desktop/bioinfo/RESULTS_LOCAL/test.bam \"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mconvertSAMtoBAM\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minputf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moutputf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'convertSAMtoBAM' is not defined"
     ]
    }
   ],
   "source": [
    "inputf = \"/Users/jenniferureta/Desktop/bioinfo/RESULTS_LOCAL/HIV2_171.sam \"\n",
    "outputf = \"/Users/jenniferureta/Desktop/bioinfo/RESULTS_LOCAL/test.bam \"\n",
    "convertSAMtoBAM(inputf, outputf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
